pkg/dynchan/pipechan.go

```diff
@@ -0,0 +1,74 @@
+/*
+ * Minio Client (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dynchan
+
+// Dynamically sized logical channel: a pipe which never blocks even when
+// it receives too many elements. Memory consumption is increased and decresed
+// on the fly following the number of elements.
+func PipeChan(capacity int) (inputCh, outputCh chan interface{}) {
```

Comment:
why take capacity? if its going to increase in size anyways. Perhaps i am seeing that you want it something like `bytes.NewBuffer([capacity]byte)` .